src/python_testing/TC_OPSTATE_2_1.py

```diff
@@ -0,0 +1,181 @@
+        cluster = Clusters.Objects.OperationalState
+        return await self.read_single_attribute_check_success(endpoint=endpoint, cluster=cluster, attribute=attribute)
+
+    async def read_and_validate_opstate(self, step, expected_state):
+        self.print_step(step, "Read OperationalState attribute")
+        operational_state = await self.read_mod_attribute_expect_success(
+            endpoint=self.endpoint, attribute=Clusters.OperationalState.Attributes.OperationalState)
+        logging.info("OperationalState: %s" % (operational_state))
+        asserts.assert_equal(operational_state, expected_state, "OperationalState should equal %s" % expected_state)
+
+    async def read_and_validate_operror(self, step, expected_error):
+        self.print_step(step, "Read OperationalError attribute")
+        operational_error = await self.read_mod_attribute_expect_success(
+            endpoint=self.endpoint, attribute=Clusters.OperationalState.Attributes.OperationalError)
+        logging.info("OperationalError: %s" % (operational_error))
+        asserts.assert_equal(operational_error.errorStateID, expected_error, "OperationalError should equal %s" % expected_error)
+
+    @async_test_body
+    async def test_TC_OPSTATE_2_1(self):
+
+        asserts.assert_true('PIXIT_ENDPOINT' in self.matter_test_config.global_test_params,
+                            "PIXIT_ENDPOINT must be included on the command line in "
+                            "the --int-arg flag as PIXIT_ENDPOINT:<endpoint>")
+
+        self.endpoint = self.matter_test_config.global_test_params['PIXIT_ENDPOINT']
+
+        attributes = Clusters.OperationalState.Attributes
+
+        self.print_step(1, "Commissioning, already done")
+
+        if self.check_pics("OPSTATE.S.A0000"):
+            self.print_step(2, "Read PhaseList attribute")
+            phase_list = await self.read_mod_attribute_expect_success(endpoint=self.endpoint, attribute=attributes.PhaseList)
+
+            if phase_list == NullValue:
+                logging.info("PhaseList is null")
+            else:
+                logging.info("PhaseList: %s" % (phase_list))
+
+                phase_list_len = len(phase_list)
+
+                asserts.assert_less_equal(phase_list_len, 32, "PhaseList must have no more than 32 entries!")
+
+        if self.check_pics("OPSTATE.S.A0001"):
+            self.print_step(3, "Read CurrentPhase attribute")
+            current_phase = await self.read_mod_attribute_expect_success(endpoint=self.endpoint, attribute=attributes.CurrentPhase)
+            logging.info("CurrentPhase: %s" % (current_phase))
+
+            if phase_list == NullValue:
+                asserts.assert_true(current_phase == NullValue, "CurrentPhase should be null")
+            else:
+                asserts.assert_true(0 <= current_phase and current_phase <= (phase_list_len - 1),
+                                    "CurrentPhase must be between 0 and (phase-list-size - 1)")
```

Comment:
would it make sense to add the current values for logging purposes? If I see this message I don't know what the current_phase is (is it negative? too large?) and I don't know what the phase_list_len is. Also the check could be using `<` and not do the -1.